File: Day6/lights.js
#!/usr/bin/env node

const FS = require("fs");

class Action {
    static fromText(text) {
        if (text.startsWith("turn on")) {
            return new TurnOn();
        }

        if (text.startsWith("turn off")) {
            return new TurnOff();
        }

        return new Toggle();
    }
}

class TurnOn extends Action {
    getValue() {
        return true;
    }
}

class TurnOff extends Action {
    getValue() {
        return false;
    }
}

class Toggle extends Action {
    getValue(value) {
        return !value;
    }
}

class Position {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}

class Instruction {
    constructor(start, end, action) {
        this.start  = start;
        this.end    = end;
        this.action = action;
    }

    static fromText(text) {
        const action  = Action.fromText(text);
        const regex   = /(\d+),(\d+) through (\d+),(\d+)/;
        const matches = text.match(regex);

        if (!matches) {
            return [];
        }

        const start = new Position(parseInt(matches[1]), parseInt(matches[2]));
        const end   = new Position(parseInt(matches[3]), parseInt(matches[4]));

        return new Instruction(start, end, action);
    }
}

class Grid {
    static get SIZE() {
        return 1000;
    }

    constructor() {
        this.grid = new Array(Grid.SIZE);

        for (let i = 0; i < Grid.SIZE; ++i) {
            this.grid[i] = new Array(Grid.SIZE);

            for (let j = 0; j < Grid.SIZE; ++j) {
                this.grid[i][j] = false;
            }
        }
    }

    update(instruction) {
        for (let x = instruction.start.x; x <= instruction.end.x; ++x) {
            for (let y = instruction.start.y; y <= instruction.end.y; ++y) {
                this.grid[x][y] = instruction.action.getValue(this.grid[x][y]);
            }
        }
    }

    count() {
        return this.grid.reduce((sum, row) => {
            return sum + row.reduce((sum, cell) => {
                return sum + (cell === true);
            }, 0);
        }, 0);
    }
}

FS.readFile("input", "utf8", function(error, contents) {
    if (error) {
        return;
    }

    const lines = contents.split("\n").filter(line => {
        return line !== "";
    });

    const instructions = lines.map(line => {
        return Instruction.fromText(line);
    });

    const grid = new Grid();

    instructions.forEach(instruction => {
        grid.update(instruction);
    });

    console.log(grid.count());
});

File: Day6/lights.hs
import Data.List.Split
import Data.List
import Data.Array

-- This solution seems correct, but doesnt work due to extreme memory consumption.

data Action = TurnOn | TurnOff | Toggle deriving (Show)

data Instruction = Instruction {
    position :: (Int, Int),
    action   :: Action
} deriving (Show)

action2value :: Action -> Int -> Int
action2value TurnOn  _ = 1
action2value TurnOff _ = 0
action2value Toggle  b = case b of 0 -> 1
                                   1 -> 0

str2action :: String -> Action
str2action str = if "turn on" `isPrefixOf` str then TurnOn
                 else if "turn off" `isPrefixOf` str then TurnOff
                 else Toggle

str2range :: String -> [ (Int, Int) ]
str2range str = range (start, end)
    where wrds  = reverse $ words $ str
          start = str2ints $ wrds !! 2
          end   = str2ints $ wrds !! 0

str2ints :: String -> (Int, Int)
str2ints str = (read x, read y)
    where [x, y] = splitOn delim str
          delim  = ","

str2instructions :: String -> [ Instruction ]
str2instructions str = [ Instruction { position = pos, action = act } | pos <- poss ]
    where poss = str2range str
          act  = str2action str

updateGrid :: Array (Int, Int) Int -> Instruction -> Array (Int, Int) Int
updateGrid arr instr = arr // [ (pos, val) ]
    where pos = position instr
          val = action2value (action instr) (arr ! pos)

main = do
    contents <- readFile "input"

    let instructions = intercalate [] $ map str2instructions (lines contents)
    let grid = array ((0,0),(999,999)) [ ((x, y), 0) | y <- [0..999], x <- [0..999] ]

    let gridOut = foldl updateGrid grid instructions

    let suma = foldl (+) 0 gridOut

    print $ suma

